compiler/optimizer/VPHandlers.cpp

```diff
@@ -9829,33 +9807,33 @@ static TR::Node *constrainIfcmpeqne(OMR::ValuePropagation *vp, TR::Node *node, b
                                                    objectClass, cpIndexOrVftSlot, symRef->getOwningMethod(vp->comp()),
                                                    vp->comp(), false, useGetResolvedInterfaceMethod);
 
-        TR_ScratchList<TR_PersistentClassInfo> subClasses(vp->comp()->trMemory());
-
-        //in AOT or when CHOpts are disabled findClassInfoAfterLocking returns NULL unconditionally
-        //we should treat this case conservatively meaning we assume that there are more than two concrete classes in the given hierarchy
-        //most likely rvm will also be NULL in this case, but it is better to not rely on this implicit connection between different
-        //CH queries.
-        TR_PersistentClassInfo *objectClassInfo = vp->comp()->getPersistentInfo()->getPersistentCHTable()->findClassInfoAfterLocking(objectClass, vp->comp());
+                TR_ScratchList<TR_PersistentClassInfo> subClasses(vp->comp()->trMemory());
 
-        bool searchSucceeded = false;
+                //in AOT or when CHOpts are disabled findClassInfoAfterLocking returns NULL unconditionally
+                //we should treat this case conservatively meaning we assume that there are more than two concrete classes in the given hierarchy
+                //most likely rvm will also be NULL in this case, but it is better to not rely on this implicit connection between different
+                //CH queries.
+                TR_PersistentClassInfo *objectClassInfo = vp->comp()->getPersistentInfo()->getPersistentCHTable()->findClassInfoAfterLocking(objectClass, vp->comp());
 
-        if (objectClassInfo)
-          {
-          TR_ClassQueries::collectAllSubClasses(objectClassInfo, &subClasses, vp->comp());
-          subClasses.add(objectClassInfo);
-          searchSucceeded = true;
-          }
+                bool searchSucceeded = false;
 
-      /*
-        the limited version of a transformation is enabled
-        Namely, we only do a transformation if there is no more than one instantiable class in the hierarchy
-        starting from objectClass. If there is more than one concrete class in hierarchy
-        overriding methods in objectClass and if any call was devirtualized by invariantargumentpreexisence downstream
-        we will produce the functionally incorrect code by swapping a profiled test with a method test as
-        the method test is insufficient for the devirtualizations done by invariantargumentpreexisence
-      */
+                if (objectClassInfo)
```

Comment:
Some suggestions to reorganize the code, unrelated to JITServer:
If `rvm` is null we have executed `TR_ClassQueries::collectAllSubClasses` for nothing. Thus I would put a `if (rvm)` test up front. Then, `searchSucceeded` is set immediately after `TR_ClassQueries::collectAllSubClasses()`; it may be better to eliminate the local variable `searchSucceded` and just insert the code:
```
               if (numConcreteClasses(&subClasses) < 2)
                {
                   TR::ResolvedMethodSymbol* cMethodSymbol = vp->comp()->getSymRefTab()->findOrCreateMethodSymbol(
                      symRef->getOwningMethodIndex(), -1, rvm, methodKind)->getSymbol()->castToResolvedMethodSymbol();

                   TR_VirtualGuardKind guardKind =
                      methodSymbol->isInterface() ? TR_InterfaceGuard :
                      TR::Compiler->cls.isAbstractClass(vp->comp(), objectClass) ? TR_AbstractGuard : TR_HierarchyGuard;

                   addDelayedConvertedGuard(node, callNode, cMethodSymbol, vGuard, vp, guardKind, TR_VftTest, objectClass);
                }
```
afterwards

Reply:
Thanks, please take a look at the lastest commit and see if it makes sense.